0101-compiling-recursive-types.md

```diff
@@ -0,0 +1,367 @@
+one form as normal when they are unified. This is implemented today as an
+operation called "fixpoint fixing" (adjusting the fixpoint of two recursive
+types to be equivalent). However, this operation is insufficient when types are
+exported across modules. When types are exported across modules, their
+relationship to types in the source module is lost. This means that a type
+exported from module A to module B may be fixpoint-fixed in module B, and end up
+in a form that is equivalent, but not equal, to its form in module A. When
+module A and B are compiled, that type will lower to equivalent but not equal
+layouts.
+
+At the time fixpoint fixing was implemented, this limitation was entirely
+overlooked. I hope it is clear that the only way fixpoint fixing can resolve
+this issue in the presence of modules is to break module isolation, and have
+typechecking be performed globally, with sharing across all modules. This has a
+number of problems, including making incremental compilation nearly impossible.
+I am happy to elaborate more as desired.
+
+The alternative is to consider a language-feature restriction that forces
+equivalent recursive types to be equal. The simplest one I can think of is what
+most other languages do - require recursive types to be nominal. In Roc, this
+would mean that recursive types must be named as opaque types.
+
+If a structural type is determined to be recursive, but has no recursion point
+that is an opaque type, that will now be an error. Otherwise, two recursive
+types can only be equivalent if they are the same opaque type - which
+necessarily means they are equal, since opaque types can only have one
+definition.
+
+#### User implications
+
+The main implication of this for Roc developers is that all recursive types will
+now have to named as opaque types explicitly.
+
+For example, the following function which calculates the length of a linked
+list, would no longer typecheck:
+
+```roc
+length = \list ->
+   when list is
+      Nil -> 0
+      Cons _ rest -> 1 + length rest
+```
+
+An appropriate error message for this program would be something like
+
+```
+--- ANONYMOUS RECURSIVE TYPE ---
+
+This function contains a recursive type that has no name:
+
+length = \list ->
+^^^^^^
+   when list is
+      Nil -> 0
+      Cons _ rest -> 1 + length rest
+
+`length` has a type that looks like
+  
+  [Nil, Cons a <RecursiveType>] as <RecursiveType> -> Num *
+
+Because `<RecursiveType>` is recursive, it must be defined as an opaque type.
+Consider defining
+
+  RecursiveType a := [Nil, Cons a (RecursiveType a)]
+
+and changing the implementation of `length` to have type
+
+  RecursiveType a -> Num *
+```
+
+Please defer discussion of the exact error message's wording to the implementation.
+See below on how to implement this error message.
+
+A suitable re-implementation of this function would be
+
+```roc
+LinkedList a := [Nil, Cons a (RecursiveType a)]
```

Comment:
```suggestion
LinkedList a := [Nil, Cons a (LinkedList a)]
```

Reply:
Thanks! I should tried to plug all this into the compiler 😅